server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -525,19 +525,22 @@ static void checkForIndexCompatibility(Logger logger, DataPath... dataPaths) thr
         logger.info("oldest index version recorded in NodeMetadata {}", metadata.oldestIndexVersion());
 
         if (metadata.oldestIndexVersion().isLegacyIndexVersion()) {
+            String bestDowngradeVersion = Build.isIndexCompatibleWithCurrentAllowingLegacy(metadata.previousNodeVersion().toString())
```

Comment:
I wonder if a call to `isIndexCompatibleWithCurrentAllowingLegacy` is really needed.
isLegacyIndexVersion() should already tell us it is Version (and in fact we are passing down `metadata.previousNodeVersion().toString()`).
Wouldn't it be better to just have the Version logic here? (`metadata.previousNodeVersion()..onOrAfter(Version.CURRENT.minimumCompatibilityVersion())`)

Reply:
That's a good point. I think we should return what's most useful to an on-prem user, because on-disk index compatibility will be less of an issue in environments that we're managing. This message should give the user what they need to figure out which Elasticsearch artifact needs to be installed, and I don't think an IndexVersion would be helpful there.

Reply:
I did some thinking, and I believe I understand better what the original version maximum check was doing. In the case where we have an 8.x node starting up, an incompatible index would be from 6.x or earlier. We want to tell the user to downgrade to 7.17.x. If the last installation was from 7.16.x or earlier, we can tell them to go to minimum wire compat version (7.17.0). Otherwise, we tell them to go to the version they were previously using, for example, 7.17.5.

The minimum index version is at the _start_ of the last major, meaning 7.0.0 in this case. We definitely don't want to tell the user to downgrade to that!

Reply:
I don't know why I'm having such a hard time thinking through what will happen with this code when breaking changes happen. It's in a context that's very specific to the 7.x -> 8.x upgrade, but who's to say it won't get re-used elsewhere.

I really don't like the idea of making the user go cross-reference things in documentation when we have a way of providing that information directly. If I were trying to figure out why my upgrade failed, I'd much prefer "go back to 7.17.5" (for which I might already have a distribution file ready) to "go find and install the distribution for the latest bugfix version of the previous major version's latest minor version."

I think I'm just struggling with how hard it's becoming to express a reference to a user-facing release in terms that would be clear to someone who just needs to figure out what version to download.

Reply:
> I think I'm just struggling with how hard it's becoming to express a reference to a user-facing release in terms that would be clear to someone who just needs to figure out what version to download.

I think this is one major drawback of removing Version: it becomes really difficult to create useful messages for the user with the correct version information. Without a concept of "previous", or any link between IndexVersion/TransportVersion and downloadable artifacts, these lines becomes really difficult to create. 
What about hardcoding? `Build.indexCompatibilityDowngradeVersion()`, always returing "7.17.x" on on-prem (and the usual hash in serverless - which BTW will generate a useless message?)

Reply:
How about: if `version` matches regex `^7\.17\.[0-9]+$`, output `version`, otherwise return `"7.17.0"` (or `7.17.x`)?

Reply:
I think using the regex is a good compromise! Consider adding an assertion too (to fail if it matches `9.`?), so when the time comes and if we need to, we will be able to update it quickly.